File: julia/src/action/ActionAst.jl
struct ActionSourceSpan
    start::Int
    stop::Int
end

abstract type ActionNode end
abstract type ActionExpr <: ActionNode end
abstract type ActionArgument end
abstract type ActionAccessSegment end

struct ActionStatement <: ActionNode
    kind::String
    source::String
    source_span::ActionSourceSpan
    expr::ActionExpr
    drops_value::Bool
end

function ActionStatement(; source, source_span, expr, drops_value = true)
    return ActionStatement("action_stmt", String(source), source_span, expr, drops_value)
end

struct ActionBlock <: ActionNode
    kind::String
    source::String
    source_span::ActionSourceSpan
    statements::Vector{ActionStatement}
end

function ActionBlock(; source, source_span, statements)
    return ActionBlock("action_block", String(source), source_span, ActionStatement[statements...])
end

struct ActionPositionalArgument <: ActionArgument
    value::ActionExpr
end

struct ActionKeywordArgument <: ActionArgument
    name::String
    value::ActionExpr
end

function ActionKeywordArgument(; name, value)
    return ActionKeywordArgument(String(name), value)
end

struct ActionCallExpr <: ActionExpr
    kind::String
    source::String
    source_span::ActionSourceSpan
    name::String
    args::Vector{ActionArgument}
    source_method::Union{Nothing,String}
    trailing_block_arg::Bool
    trailing_block_source_span::Union{Nothing,ActionSourceSpan}
end

function ActionCallExpr(;
    source,
    source_span,
    name,
    args,
    source_method = nothing,
    trailing_block_arg = false,
    trailing_block_source_span = nothing,
)
    return ActionCallExpr(
        "call",
        String(source),
        source_span,
        String(name),
        ActionArgument[args...],
        source_method === nothing ? nothing : String(source_method),
        trailing_block_arg,
        trailing_block_source_span,
    )
end

struct ActionVariableExpr <: ActionExpr
    kind::String
    source::String
    source_span::ActionSourceSpan
    name::String
end

function ActionVariableExpr(; source, source_span, name)
    return ActionVariableExpr("variable", String(source), source_span, String(name))
end

struct ActionIndexedVarExpr <: ActionExpr
    kind::String
    source::String
    source_span::ActionSourceSpan
    name::String
    index::ActionExpr
end

function ActionIndexedVarExpr(; source, source_span, name, index)
    return ActionIndexedVarExpr("indexed_var", String(source), source_span, String(name), index)
end

struct ActionKeyAccessSegment <: ActionAccessSegment
    kind::String
    value::String
    source::String
    source_span::ActionSourceSpan
end

function ActionKeyAccessSegment(; value, source, source_span)
    return ActionKeyAccessSegment("key", String(value), String(source), source_span)
end

struct ActionIndexAccessSegment <: ActionAccessSegment
    kind::String
    expr::ActionExpr
    source::String
    source_span::ActionSourceSpan
end

function ActionIndexAccessSegment(; expr, source, source_span)
    return ActionIndexAccessSegment("index", expr, String(source), source_span)
end

struct ActionNestedAccessExpr <: ActionExpr
    kind::String
    source::String
    source_span::ActionSourceSpan
    base::String
    segments::Vector{ActionAccessSegment}
end

function ActionNestedAccessExpr(; source, source_span, base, segments)
    return ActionNestedAccessExpr(
        "nested_access",
        String(source),
        source_span,
        String(base),
        ActionAccessSegment[segments...],
    )
end

struct ActionArrayLiteralExpr <: ActionExpr
    kind::String
    source::String
    source_span::ActionSourceSpan
    items::Vector{ActionExpr}
end

function ActionArrayLiteralExpr(; source, source_span, items)
    return ActionArrayLiteralExpr("array_literal", String(source), source_span, ActionExpr[items...])
end

struct ActionHashLiteralEntry
    key::ActionExpr
    value::ActionExpr
end

function ActionHashLiteralEntry(; key, value)
    return ActionHashLiteralEntry(key, value)
end

struct ActionHashLiteralExpr <: ActionExpr
    kind::String
    source::String
    source_span::ActionSourceSpan
    entries::Vector{ActionHashLiteralEntry}
end

function ActionHashLiteralExpr(; source, source_span, entries)
    return ActionHashLiteralExpr(
        "hash_literal",
        String(source),
        source_span,
        ActionHashLiteralEntry[entries...],
    )
end

struct ActionBlockValueExpr <: ActionExpr
    kind::String
    source::String
    source_span::ActionSourceSpan
    block::ActionBlock
end

function ActionBlockValueExpr(; source, source_span, block)
    return ActionBlockValueExpr("block_value", String(source), source_span, block)
end

struct ActionStringLiteralExpr <: ActionExpr
    kind::String
    source::String
    source_span::ActionSourceSpan
    value::String
    quote_char::String
end

function ActionStringLiteralExpr(; source, source_span, value, quote_char)
    return ActionStringLiteralExpr("string", String(source), source_span, String(value), String(quote_char))
end

struct ActionNumberLiteralExpr <: ActionExpr
    kind::String
    source::String
    source_span::ActionSourceSpan
    value::Union{Int,Float64}
end

function ActionNumberLiteralExpr(; source, source_span, value)
    return ActionNumberLiteralExpr("number", String(source), source_span, value)
end

struct ActionBooleanLiteralExpr <: ActionExpr
    kind::String
    source::String
    source_span::ActionSourceSpan
    value::Bool
end

function ActionBooleanLiteralExpr(; source, source_span, value)
    return ActionBooleanLiteralExpr("boolean", String(source), source_span, value)
end

struct ActionRegexLiteralExpr <: ActionExpr
    kind::String
    source::String
    source_span::ActionSourceSpan
    pattern::String
    flags::String
end

function ActionRegexLiteralExpr(; source, source_span, pattern, flags = "")
    return ActionRegexLiteralExpr("regex", String(source), source_span, String(pattern), String(flags))
end

struct ActionUndefExpr <: ActionExpr
    kind::String
    source::String
    source_span::ActionSourceSpan
end

function ActionUndefExpr(; source, source_span)
    return ActionUndefExpr("undef", String(source), source_span)
end

struct ActionAssignScalarExpr <: ActionExpr
    kind::String
    source::String
    source_span::ActionSourceSpan
    name::String
    value::ActionExpr
end

function ActionAssignScalarExpr(; source, source_span, name, value)
    return ActionAssignScalarExpr("assign_scalar", String(source), source_span, String(name), value)
end

struct ActionAssignArrayAppendExpr <: ActionExpr
    kind::String
    source::String
    source_span::ActionSourceSpan
    name::String
    value::ActionExpr
end

function ActionAssignArrayAppendExpr(; source, source_span, name, value)
    return ActionAssignArrayAppendExpr("assign_array_append", String(source), source_span, String(name), value)
end

struct ActionAssignHashIndexExpr <: ActionExpr
    kind::String
    source::String
    source_span::ActionSourceSpan
    name::String
    key::ActionExpr
    value::ActionExpr
end

function ActionAssignHashIndexExpr(; source, source_span, name, key, value)
    return ActionAssignHashIndexExpr("assign_hash_index", String(source), source_span, String(name), key, value)
end

struct ActionAssignNestedAccessExpr <: ActionExpr
    kind::String
    source::String
    source_span::ActionSourceSpan
    base::String
    segments::Vector{ActionAccessSegment}
    value::ActionExpr
end

function ActionAssignNestedAccessExpr(; source, source_span, base, segments, value)
    return ActionAssignNestedAccessExpr(
        "assign_nested_access",
        String(source),
        source_span,
        String(base),
        ActionAccessSegment[segments...],
        value,
    )
end

struct ActionFluentCall
    method::String
    args::Vector{ActionArgument}
    source::String
    source_span::ActionSourceSpan
    source_method::Union{Nothing,String}
    trailing_block_arg::Bool
    receiver_trailing_block_arg::Bool
    trailing_block_source_span::Union{Nothing,ActionSourceSpan}
end

function ActionFluentCall(;
    method,
    args,
    source,
    source_span,
    source_method = nothing,
    trailing_block_arg = false,
    receiver_trailing_block_arg = false,
    trailing_block_source_span = nothing,
)
    return ActionFluentCall(
        String(method),
        ActionArgument[args...],
        String(source),
        source_span,
        source_method === nothing ? nothing : String(source_method),
        trailing_block_arg,
        receiver_trailing_block_arg,
        trailing_block_source_span,
    )
end

struct ActionFluentChainExpr <: ActionExpr
    kind::String
    source::String
    source_span::ActionSourceSpan
    receiver::ActionExpr
    calls::Vector{ActionFluentCall}
end

function ActionFluentChainExpr(; source, source_span, receiver, calls)
    return ActionFluentChainExpr(
        "fluent_chain",
        String(source),
        source_span,
        receiver,
        ActionFluentCall[calls...],
    )
end

struct ActionControlIfExpr <: ActionExpr
    kind::String
    source::String
    source_span::ActionSourceSpan
    keyword::String
    canonical_keyword::String
    branch_role::String
    condition::ActionExpr
    args::Vector{ActionArgument}
    body::Union{Nothing,ActionBlock}
    body_source_span::Union{Nothing,ActionSourceSpan}
end

function ActionControlIfExpr(;
    source,
    source_span,
    keyword,
    canonical_keyword,
    branch_role,
    condition,
    args,
    body = nothing,
    body_source_span = nothing,
)
    return ActionControlIfExpr(
        "control_if",
        String(source),
        source_span,
        String(keyword),
        String(canonical_keyword),
        String(branch_role),
        condition,
        ActionArgument[args...],
        body,
        body_source_span,
    )
end

struct ActionControlElseExpr <: ActionExpr
    kind::String
    source::String
    source_span::ActionSourceSpan
    keyword::String
    canonical_keyword::String
    args::Vector{ActionArgument}
    body::Union{Nothing,ActionBlock}
    body_source_span::Union{Nothing,ActionSourceSpan}
end

function ActionControlElseExpr(;
    source,
    source_span,
    keyword,
    canonical_keyword,
    args,
    body = nothing,
    body_source_span = nothing,
)
    return ActionControlElseExpr(
        "control_else",
        String(source),
        source_span,
        String(keyword),
        String(canonical_keyword),
        ActionArgument[args...],
        body,
        body_source_span,
    )
end

struct ActionControlMarkerExpr <: ActionExpr
    kind::String
    source::String
    source_span::ActionSourceSpan
    keyword::String
    canonical_keyword::String
    args::Vector{ActionArgument}
end

function ActionControlMarkerExpr(; kind, source, source_span, keyword, canonical_keyword, args)
    return ActionControlMarkerExpr(
        String(kind),
        String(source),
        source_span,
        String(keyword),
        String(canonical_keyword),
        ActionArgument[args...],
    )
end

struct ActionControlWhileExpr <: ActionExpr
    kind::String
    source::String
    source_span::ActionSourceSpan
    keyword::String
    canonical_keyword::String
    condition::ActionExpr
    args::Vector{ActionArgument}
    body::Union{Nothing,ActionBlock}
    body_source_span::Union{Nothing,ActionSourceSpan}
end

function ActionControlWhileExpr(;
    source,
    source_span,
    keyword,
    canonical_keyword,
    condition,
    args,
    body = nothing,
    body_source_span = nothing,
)
    return ActionControlWhileExpr(
        "control_while",
        String(source),
        source_span,
        String(keyword),
        String(canonical_keyword),
        condition,
        ActionArgument[args...],
        body,
        body_source_span,
    )
end

struct ActionControlSwitchExpr <: ActionExpr
    kind::String
    source::String
    source_span::ActionSourceSpan
    keyword::String
    canonical_keyword::String
    source_expr::ActionExpr
    args::Vector{ActionArgument}
    body::Union{Nothing,ActionBlock}
    body_source_span::Union{Nothing,ActionSourceSpan}
    cases::Vector{ActionExpr}
    default_case::Union{Nothing,ActionExpr}
end

function ActionControlSwitchExpr(;
    source,
    source_span,
    keyword,
    canonical_keyword,
    source_expr,
    args,
    body = nothing,
    body_source_span = nothing,
    cases = Any[],
    default_case = nothing,
)
    return ActionControlSwitchExpr(
        "control_switch",
        String(source),
        source_span,
        String(keyword),
        String(canonical_keyword),
        source_expr,
        ActionArgument[args...],
        body,
        body_source_span,
        ActionExpr[cases...],
        default_case,
    )
end

struct ActionControlCaseExpr <: ActionExpr
    kind::String
    source::String
    source_span::ActionSourceSpan
    keyword::String
    canonical_keyword::String
    match::ActionExpr
    args::Vector{ActionArgument}
    body::Union{Nothing,ActionBlock}
    body_source_span::Union{Nothing,ActionSourceSpan}
end

function ActionControlCaseExpr(;
    source,
    source_span,
    keyword,
    canonical_keyword,
    match,
    args,
    body = nothing,
    body_source_span = nothing,
)
    return ActionControlCaseExpr(
        "control_case",
        String(source),
        source_span,
        String(keyword),
        String(canonical_keyword),
        match,
        ActionArgument[args...],
        body,
        body_source_span,
    )
end

struct ActionControlDefaultExpr <: ActionExpr
    kind::String
    source::String
    source_span::ActionSourceSpan
    keyword::String
    canonical_keyword::String
    args::Vector{ActionArgument}
    body::Union{Nothing,ActionBlock}
    body_source_span::Union{Nothing,ActionSourceSpan}
end

function ActionControlDefaultExpr(;
    source,
    source_span,
    keyword,
    canonical_keyword,
    args,
    body = nothing,
    body_source_span = nothing,
)
    return ActionControlDefaultExpr(
        "control_default",
        String(source),
        source_span,
        String(keyword),
        String(canonical_keyword),
        ActionArgument[args...],
        body,
        body_source_span,
    )
end

struct ActionRawExpr <: ActionExpr
    kind::String
    source::String
    source_span::ActionSourceSpan
    reason::String
end

function ActionRawExpr(; source, source_span, reason)
    return ActionRawExpr("raw_perl", String(source), source_span, String(reason))
end

to_json(span::ActionSourceSpan) = Dict("start" => span.start, "end" => span.stop)

function _action_base_json(node)
    return Dict{String,Any}(
        "kind" => node.kind,
        "source" => node.source,
        "source_span" => to_json(node.source_span),
    )
end

function to_json(block::ActionBlock)
    result = _action_base_json(block)
    result["statements"] = [to_json(statement) for statement in block.statements]
    return result
end

function to_json(statement::ActionStatement)
    result = _action_base_json(statement)
    result["expr"] = to_json(statement.expr)
    result["drops_value"] = statement.drops_value
    return result
end

to_json(argument::ActionPositionalArgument) = to_json(argument.value)
to_json(argument::ActionKeywordArgument) = Dict("name" => argument.name, "value" => to_json(argument.value))

function to_json(expr::ActionCallExpr)
    result = _action_base_json(expr)
    result["name"] = expr.name
    _put_if_present!(result, "source_method", expr.source_method)
    result["args"] = [to_json(argument) for argument in expr.args]
    if expr.trailing_block_arg
        result["trailing_block_arg"] = true
    end
    if expr.trailing_block_source_span !== nothing
        result["trailing_block_source_span"] = to_json(expr.trailing_block_source_span)
    end
    return result
end

function to_json(expr::ActionVariableExpr)
    result = _action_base_json(expr)
    result["name"] = expr.name
    return result
end

function to_json(expr::ActionIndexedVarExpr)
    result = _action_base_json(expr)
    result["name"] = expr.name
    result["index"] = to_json(expr.index)
    return result
end

function to_json(segment::ActionKeyAccessSegment)
    return Dict(
        "kind" => segment.kind,
        "value" => segment.value,
        "source" => segment.source,
        "source_span" => to_json(segment.source_span),
    )
end

function to_json(segment::ActionIndexAccessSegment)
    return Dict(
        "kind" => segment.kind,
        "expr" => to_json(segment.expr),
        "source" => segment.source,
        "source_span" => to_json(segment.source_span),
    )
end

function to_json(expr::ActionNestedAccessExpr)
    result = _action_base_json(expr)
    result["base"] = expr.base
    result["segments"] = [to_json(segment) for segment in expr.segments]
    return result
end

function to_json(expr::ActionArrayLiteralExpr)
    result = _action_base_json(expr)
    result["items"] = [to_json(item) for item in expr.items]
    return result
end

to_json(entry::ActionHashLiteralEntry) = Dict("key" => to_json(entry.key), "value" => to_json(entry.value))

function to_json(expr::ActionHashLiteralExpr)
    result = _action_base_json(expr)
    result["entries"] = [to_json(entry) for entry in expr.entries]
    return result
end

function to_json(expr::ActionBlockValueExpr)
    result = _action_base_json(expr)
    result["block"] = to_json(expr.block)
    return result
end

function to_json(expr::ActionStringLiteralExpr)
    result = _action_base_json(expr)
    result["value"] = expr.value
    result["quote"] = expr.quote_char
    return result
end

function to_json(expr::ActionNumberLiteralExpr)
    result = _action_base_json(expr)
    result["value"] = expr.value
    return result
end

function to_json(expr::ActionBooleanLiteralExpr)
    result = _action_base_json(expr)
    result["value"] = expr.value
    return result
end

function to_json(expr::ActionRegexLiteralExpr)
    result = _action_base_json(expr)
    result["pattern"] = expr.pattern
    result["flags"] = expr.flags
    return result
end

to_json(expr::ActionUndefExpr) = _action_base_json(expr)

function to_json(expr::ActionAssignScalarExpr)
    result = _action_base_json(expr)
    result["name"] = expr.name
    result["value"] = to_json(expr.value)
    return result
end

function to_json(expr::ActionAssignArrayAppendExpr)
    result = _action_base_json(expr)
    result["name"] = expr.name
    result["value"] = to_json(expr.value)
    return result
end

function to_json(expr::ActionAssignHashIndexExpr)
    result = _action_base_json(expr)
    result["name"] = expr.name
    result["key"] = to_json(expr.key)
    result["value"] = to_json(expr.value)
    return result
end

function to_json(expr::ActionAssignNestedAccessExpr)
    result = _action_base_json(expr)
    result["base"] = expr.base
    result["segments"] = [to_json(segment) for segment in expr.segments]
    result["value"] = to_json(expr.value)
    return result
end

function to_json(call::ActionFluentCall)
    result = Dict{String,Any}(
        "method" => call.method,
        "args" => [to_json(argument) for argument in call.args],
        "source" => call.source,
        "source_span" => to_json(call.source_span),
    )
    _put_if_present!(result, "source_method", call.source_method)
    if call.trailing_block_arg
        result["trailing_block_arg"] = true
    end
    if call.receiver_trailing_block_arg
        result["receiver_trailing_block_arg"] = true
    end
    if call.trailing_block_source_span !== nothing
        result["trailing_block_source_span"] = to_json(call.trailing_block_source_span)
    end
    return result
end

function to_json(expr::ActionFluentChainExpr)
    result = _action_base_json(expr)
    result["receiver"] = to_json(expr.receiver)
    result["calls"] = [to_json(call) for call in expr.calls]
    return result
end

function to_json(expr::ActionControlIfExpr)
    result = _action_base_json(expr)
    result["keyword"] = expr.keyword
    result["canonical_keyword"] = expr.canonical_keyword
    result["branch_role"] = expr.branch_role
    result["condition"] = to_json(expr.condition)
    result["args"] = [to_json(argument) for argument in expr.args]
    _put_action_block_fields!(result, expr.body, expr.body_source_span)
    return result
end

function to_json(expr::ActionControlElseExpr)
    result = _action_base_json(expr)
    result["keyword"] = expr.keyword
    result["canonical_keyword"] = expr.canonical_keyword
    result["branch_role"] = "else"
    result["args"] = [to_json(argument) for argument in expr.args]
    _put_action_block_fields!(result, expr.body, expr.body_source_span)
    return result
end

function to_json(expr::ActionControlMarkerExpr)
    result = _action_base_json(expr)
    result["keyword"] = expr.keyword
    result["canonical_keyword"] = expr.canonical_keyword
    result["args"] = [to_json(argument) for argument in expr.args]
    return result
end

function to_json(expr::ActionControlWhileExpr)
    result = _action_base_json(expr)
    result["keyword"] = expr.keyword
    result["canonical_keyword"] = expr.canonical_keyword
    result["condition"] = to_json(expr.condition)
    result["args"] = [to_json(argument) for argument in expr.args]
    _put_action_block_fields!(result, expr.body, expr.body_source_span)
    return result
end

function to_json(expr::ActionControlSwitchExpr)
    result = _action_base_json(expr)
    result["keyword"] = expr.keyword
    result["canonical_keyword"] = expr.canonical_keyword
    result["source_expr"] = to_json(expr.source_expr)
    result["args"] = [to_json(argument) for argument in expr.args]
    _put_action_block_fields!(result, expr.body, expr.body_source_span)
    if !isempty(expr.cases)
        result["cases"] = [to_json(item) for item in expr.cases]
    end
    if expr.default_case !== nothing
        result["default"] = to_json(expr.default_case)
    end
    return result
end

function to_json(expr::ActionControlCaseExpr)
    result = _action_base_json(expr)
    result["keyword"] = expr.keyword
    result["canonical_keyword"] = expr.canonical_keyword
    result["match"] = to_json(expr.match)
    result["args"] = [to_json(argument) for argument in expr.args]
    _put_action_block_fields!(result, expr.body, expr.body_source_span)
    return result
end

function to_json(expr::ActionControlDefaultExpr)
    result = _action_base_json(expr)
    result["keyword"] = expr.keyword
    result["canonical_keyword"] = expr.canonical_keyword
    result["args"] = [to_json(argument) for argument in expr.args]
    _put_action_block_fields!(result, expr.body, expr.body_source_span)
    return result
end

function to_json(expr::ActionRawExpr)
    result = _action_base_json(expr)
    result["reason"] = expr.reason
    return result
end

function _put_action_block_fields!(result, body, body_source_span)
    if body !== nothing
        result["body"] = to_json(body)
    end
    if body_source_span !== nothing
        result["body_source_span"] = to_json(body_source_span)
    end
    return result
end

File: lua/src/linkedspec/action_ast.lua
local json = require("linkedspec.json")

local M = {}

local NODE_MTS = {}

local function metatable_for(node_type)
  local metatable = NODE_MTS[node_type]
  if not metatable then
    metatable = { __action_ast_type = node_type }
    NODE_MTS[node_type] = metatable
  end
  return metatable
end

local function node(node_type, fields)
  return setmetatable(fields, metatable_for(node_type))
end

function M.node_type(value)
  if type(value) ~= "table" then
    return nil
  end
  local metatable = getmetatable(value)
  return metatable and metatable.__action_ast_type or nil
end

function M.source_span(start_position, end_position)
  return node("ActionSourceSpan", { start = start_position, ["end"] = end_position })
end

function M.block(source, source_span, statements)
  return node("ActionBlock", {
    kind = "action_block",
    source = source,
    source_span = source_span,
    statements = statements,
  })
end

function M.statement(source, source_span, expr, drops_value)
  return node("ActionStatement", {
    kind = "action_stmt",
    source = source,
    source_span = source_span,
    expr = expr,
    drops_value = drops_value ~= false,
  })
end

function M.expr(kind, source, source_span, fields)
  fields = fields or {}
  fields.kind = kind
  fields.source = source
  fields.source_span = source_span
  return node("ActionExpr", fields)
end

function M.staged_parse_job_options(fields)
  return node("ActionStagedParseJobOptions", fields)
end

function M.staged_parse_job_direct_text_plan(source, index)
  local fields = { kind = "direct_span", source = source }
  if index ~= nil then fields.index = index end
  return node("ActionStagedParseJobDirectTextPlan", fields)
end

function M.staged_parse_job_derived_text_plan(segments)
  return node("ActionStagedParseJobDerivedTextPlan", {
    kind = "derived_text",
    policy = "concatenate_in_order",
    segments = segments,
  })
end

function M.positional_argument(value)
  return node("ActionArgument", { argument_kind = "positional", value = value })
end

function M.keyword_argument(name, value)
  return node("ActionArgument", { argument_kind = "keyword", name = name, value = value })
end

function M.access_segment(kind, source, source_span, fields)
  fields = fields or {}
  fields.kind = kind
  fields.source = source
  fields.source_span = source_span
  return node("ActionAccessSegment", fields)
end

function M.write_path_segment(source, source_span, expression)
  return node("ActionWritePathSegment", {
    kind = "path_segment",
    source = source,
    source_span = source_span,
    expression = expression,
  })
end

function M.receiver_mutation_binding_reference(name, source, source_span)
  return node("ActionReceiverMutationBindingReference", {
    kind = "binding_reference",
    name = name,
    source = source,
    source_span = source_span,
  })
end

function M.receiver_mutation_callback(source, source_span, body)
  return node("ActionReceiverMutationCallback", {
    kind = "block_value",
    source = source,
    source_span = source_span,
    body = body,
  })
end

function M.receiver_mutation_call(fields)
  fields.kind = "receiver_mutation_call"
  return node("ActionReceiverMutationCall", fields)
end

function M.receiver_mutation_continuation_call(fields)
  fields.kind = "fluent_call"
  return node("ActionReceiverMutationContinuationCall", fields)
end

function M.hash_entry(key, value)
  return node("ActionHashEntry", { key = key, value = value })
end

function M.callable_signature(positional_params, rest_param)
  local positional = json.array()
  for index, name in ipairs(positional_params or {}) do positional[index] = name end
  local rest = rest_param or json.null
  return node("ActionCallableSignature", {
    kind = "callable_signature",
    version = 1,
    positional_params = positional,
    rest_param = rest,
    min_arity = #positional,
    max_arity = rest == json.null and #positional or json.null,
  })
end

function M.contextual_callable_signature()
  return M.callable_signature({}, nil)
end

function M.codeblock_literal(source, source_span, signature, body_source, body_ast, body_span)
  return M.expr("codeblock_literal", source, source_span, {
    version = 1,
    signature = signature,
    body_source = body_source,
    body_ast = body_ast,
    source_text = source,
    body_span = body_span,
  })
end

function M.codeblock_literal_error(source, source_span, code)
  return M.expr("codeblock_literal_error", source, source_span, { code = code })
end

function M.contextual_codeblock_argument(block_value)
  if M.node_type(block_value) ~= "ActionExpr" or block_value.kind ~= "block_value" or
      M.node_type(block_value.block) ~= "ActionBlock" then
    error("action AST contextual codeblock argument requires a block_value expression", 0)
  end
  return M.expr("codeblock_argument", block_value.source, block_value.source_span, {
    version = 1,
    signature = M.contextual_callable_signature(),
    body_source = block_value.block.source,
    body_ast = block_value.block,
    source_text = block_value.source,
    body_span = block_value.block.source_span,
    block = block_value.block,
  })
end

function M.fluent_call(method, args, source, source_span, fields)
  fields = fields or {}
  fields.method = method
  fields.args = args
  fields.source = source
  fields.source_span = source_span
  return node("ActionFluentCall", fields)
end

function M.removed_aggregate_selector_diagnostic(selector)
  return "aggregate_selector_removed surface=" .. selector.surface ..
    " identifier=" .. selector.identifier .. " replacement=" .. selector.identifier
end

local find_removed_aggregate_selector

local function find_in_args(args)
  for _, argument in ipairs(args or {}) do
    local selector = find_removed_aggregate_selector(argument.value)
    if selector then return selector end
  end
  return nil
end

local function find_in_block(block)
  if not block then return nil end
  for _, statement in ipairs(block.statements or {}) do
    local selector = find_removed_aggregate_selector(statement.expr)
    if selector then return selector end
  end
  return nil
end

local function find_in_segments(segments)
  for _, segment in ipairs(segments or {}) do
    if segment.kind == "index" then
      local selector = find_removed_aggregate_selector(segment.expr)
      if selector then return selector end
    elseif segment.kind == "path_segment" then
      local selector = find_removed_aggregate_selector(segment.expression)
      if selector then return selector end
    end
  end
  return nil
end

find_removed_aggregate_selector = function(value)
  if type(value) ~= "table" then return nil end
  if value.kind == "action_block" then return find_in_block(value) end

  local kind = value.kind
  if kind == "call" then
    if (value.name == "array" or value.name == "hash") and #value.args == 1 then
      local argument = value.args[1]
      if argument.argument_kind == "positional" and argument.value and argument.value.kind == "variable" then
        return { surface = value.name, identifier = argument.value.name }
      end
    end
    return find_in_args(value.args)
  elseif kind == "recognition_checkpoint" or kind == "recognize_once" or
      kind == "observe_recognition" or
      kind == "recognition_commit" or kind == "recognition_rollback" or
      kind == "progressive_dispatch_span" or kind == "staged_parse_job_marker" then
    return nil
  elseif kind == "fluent_chain" then
    local selector = find_removed_aggregate_selector(value.receiver)
    if selector then return selector end
    for _, call in ipairs(value.calls or {}) do
      selector = find_in_args(call.args)
      if selector then return selector end
    end
    return nil
  elseif kind == "receiver_mutation_chain" then
    local selector = find_in_block(value.mutation and value.mutation.callback and value.mutation.callback.body)
    if selector then return selector end
    for _, call in ipairs(value.continuation or {}) do
      selector = find_in_args(call.args)
      if selector then return selector end
    end
    return nil
  elseif kind == "assign_scalar" or kind == "assign_array_append" then
    return find_removed_aggregate_selector(value.value)
  elseif kind == "assign_hash_index" then
    return find_removed_aggregate_selector(value.key) or find_removed_aggregate_selector(value.value)
  elseif kind == "assign_nested_access" then
    return find_in_segments(value.segments) or find_removed_aggregate_selector(value.value)
  elseif kind == "indexed_var" then
    return find_removed_aggregate_selector(value.index)
  elseif kind == "nested_access" then
    return find_in_segments(value.segments)
  elseif kind == "value_access" then
    return find_removed_aggregate_selector(value.receiver) or find_in_segments(value.segments)
  elseif kind == "array_literal" then
    for _, item in ipairs(value.items or {}) do
      local selector = find_removed_aggregate_selector(item)
      if selector then return selector end
    end
    return nil
  elseif kind == "hash_literal" then
    for _, entry in ipairs(value.entries or {}) do
      local selector = find_removed_aggregate_selector(entry.key) or
        find_removed_aggregate_selector(entry.value)
      if selector then return selector end
    end
    return nil
  elseif kind == "codeblock_literal" or kind == "codeblock_literal_error" then
    return nil
  elseif kind == "block_value" or kind == "codeblock_argument" then
    return find_in_block(value.block)
  elseif kind == "control_if" or kind == "control_while" then
    return find_removed_aggregate_selector(value.condition) or find_in_args(value.args) or find_in_block(value.body)
  elseif kind == "control_switch" then
    local selector = find_removed_aggregate_selector(value.source_expr) or find_in_args(value.args) or
      find_in_block(value.body)
    if selector then return selector end
    for _, case_expr in ipairs(value.cases or {}) do
      selector = find_removed_aggregate_selector(case_expr)
      if selector then return selector end
    end
    return find_removed_aggregate_selector(value.default)
  elseif kind == "control_case" then
    return find_removed_aggregate_selector(value.match) or find_in_args(value.args) or find_in_block(value.body)
  elseif kind == "control_else" or kind == "control_default" then
    return find_in_args(value.args) or find_in_block(value.body)
  elseif kind and kind:match("^control_") then
    return find_in_args(value.args)
  end
  return nil
end

function M.find_removed_aggregate_selector(value)
  return find_removed_aggregate_selector(value)
end

local project

local function project_list(values)
  local result = json.array()
  for index, value in ipairs(values) do
    result[index] = project(value)
  end
  return result
end

project = function(value)
  local node_type = M.node_type(value)
  if node_type == nil then
    local value_type = type(value)
    if value == json.null or value_type == "string" or value_type == "number" or value_type == "boolean" then
      return value
    elseif value_type == "table" then
      return project_list(value)
    end
    error("action AST projection cannot encode " .. value_type, 0)
  end
  if node_type == "ActionArgument" and value.argument_kind == "positional" then
    return project(value.value)
  end
  if node_type == "ActionExpr" and value.kind == "codeblock_literal" then
    return json.harray({
      kind = value.kind,
      version = value.version,
      signature = project(value.signature),
      body_source = value.body_source,
      body_ast = project(value.body_ast),
      source_text = value.source_text,
      source_span = project(value.source_span),
      body_span = project(value.body_span),
    })
  end

  local result = json.harray()
  for key, item in pairs(value) do
    if key ~= "argument_kind" and item ~= nil then
      result[key] = project(item)
    end
  end
  if node_type == "ActionArgument" then
    result.name = value.name
    result.value = project(value.value)
  end
  return result
end

function M.to_json(value)
  return project(value)
end

return M
